### PR TITLE
add base64, base64url and hex encoding

### PR DESCRIPTION
**Pull request by fubhy**
Moving this to effect/platform per your suggestion @tim-smart.

Original issue: https://github.com/Effect-TS/data/issues/469

---

I think moving this to `effect/platform` makes sense if we want to use native APIs where possible. However I ran into a couple issues while doing so ... The `Buffer` APIs for base64 & hex encoding are anything but strict. Instead of throwing an error in case of invalid input, they simply skip any invalid characters and whitespace and produce garbage. I'm wondering if we should even rely on the `Buffer` based APIs for that reason? I'd much prefer strict-by-default decoding than silently producing garbage.

These are imho our options:

1. Accept that different platforms will have different behavior (I don't like this)
2. Use the same loose decoding behavior also for the platform agnostic custom implementation (I don't like this)
3. Use the custom cross-platform implementation with strict behavior everywhere (no `Buffer` usage)
4. Perform regex input validation before decoding through the`Buffer` API (performance impact?)
5. Add a `strict` flag via a second `options` parameter to the shared `decode` interface that defaults to `true`

### CODE REVIEW

**Comment by fubhy** [comment]
For consideration:

```
 BENCH  Summary

  platform - test/Encoding.bench.ts > Base64 > Encode
    1.92x faster than platform-node

  platform-node - test/Encoding.bench.ts > Base64 > Decode
    1.03x faster than platform
```

TL;DR: Enoding is consistently significantly faster with the custom algorithm. Decoding is marginally faster through the `Buffer` API.

File: packages/platform/src/internal/encoding/Base64.ts [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+  }
+
+  return result
+}
+
+/** @internal */
+export const decode = (str: string) => {
+  const length = str.length
+  if (length % 4 !== 0) {
+    throw new TypeError("Invalid base64 string")
```
Comment (tim-smart) [inline packages/platform/src/internal/encoding/Base64.ts:37]: I think the internal apis should just return Either instead of throwing.

File: packages/platform-node/src/Encoding.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+export { Base64DecodeError, Base64UrlDecodeError, HexDecodeError } from "@effect/platform/Encoding"
+
+/**
+ * Encodes a Uint8Array into a base64 (RFC4648) string.
+ *
+ * @category encoding
+ * @since 1.0.0
+ */
+export const encodeBase64 = Base64.encode
```
Comment (tim-smart) [inline packages/platform-node/src/Encoding.ts:24]: The public modules should all be annotated with types.

**Review by tim-smart — CHANGES_REQUESTED** [review]
OK after considering the lack of performance impact, having separate implementations for each platform doesn't make much sense.

So continuing this in /data is likely the better option.

**Comment by fubhy** [comment]
Ok! Continuing in /data
